### PR TITLE
QPPCT-328: cpc ecqm breakdown

### PR DESCRIPTION
**Pull request by clydet**
### Information
- JIRA story QPPCT-328

### Changes proposed in this PR:
- Introduces CPC+ measure group validation

### Checklist 
- [x] All JUnit tests pass (`mvn clean verify`).
- [x] New unit tests written to cover new functionality.
- [x] Added and updated JavaDocs for non-test classes and methods.
- [x] No local design debt. Do you feel that something is "ugly" after your changes?
- [x] Updated documentation (`README.md`, etc.) depending if the changes require it.

### CODE REVIEW

File: converter/src/main/java/gov/cms/qpp/conversion/model/validation/MeasureConfig.java [hunk 1]
```diff
@@ -31,6 +31,7 @@
 
 	@JsonProperty("eMeasureUuid")
 	private String electronicMeasureVerUuid;
+	private String cpcPlusGroup;
```
Comment (omnivion) [inline converter/src/main/java/gov/cms/qpp/conversion/model/validation/MeasureConfig.java]: does this need a @JsonProperty? Do we even need this annotation? I'm a little unfamiliar with the json pojo framework we're using.

File: converter/src/test/java/gov/cms/qpp/conversion/validate/CpcQualityMeasureSectionValidatorTest.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.stream.Stream;
+
+import static com.google.common.truth.Truth.assertThat;
+
+class CpcQualityMeasureSectionValidatorTest {
```
Comment (halprin) [inline converter/src/test/java/gov/cms/qpp/conversion/validate/CpcQualityMeasureSectionValidatorTest.java]: I'd like to see a positive test where we have a node that contains all the measures we need and verify that no validation error is created.

File: converter/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java [hunk 3]
```diff
@@ -95,7 +95,12 @@
 	QUALITY_MEASURE_ID_INCORRECT_PERFORMANCE_UUID(60, "The eCQM (electronic measure id: %s) has a %s with an "
 			+ "incorrect UUID of %s", true),
 	QUALITY_MEASURE_ID_MISSING_SINGLE_PERFORMANCE_RATE(61, "A Performance Rate must contain a single "
-			+ "Performance Rate UUID");
+			+ "Performance Rate UUID"),
+	CPC_PLUS_TOO_FEW_QUALITY_MEASURE_CATEGORY(62,
```
Comment (sfradkin) [inline converter/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java]: There will be error id collisions between the new errors added in this PR and the errors added in https://github.com/CMSgov/qpp-conversion-tool/pull/445.  Please work with @halprin to ensure that both PRs have errors with unique ids.

File: converter/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java [hunk 4]
```diff
@@ -95,7 +95,12 @@
...
+			"CPCPLUS Submissions must have at least %d of the following %s measures: %s.", true),
```
[the remark this answers is not in the document]
Comment (sfradkin) [reply converter/src/main/java/gov/cms/qpp/conversion/model/error/ErrorCode.java]: Agreed

File: converter/src/main/java/gov/cms/qpp/conversion/model/validation/MeasureConfig.java (hunk 1 above)
Comment (sfradkin) [reply converter/src/main/java/gov/cms/qpp/conversion/model/validation/MeasureConfig.java]: Usually you use `@JsonProperty` when you want the JSON emitter to output a different field name than the actual field name in the object you're serializing.  Looks like that's what's going on here, but @clydet can verify.

Comment (omnivion) [reply converter/src/main/java/gov/cms/qpp/conversion/model/validation/MeasureConfig.java]: Gotcha. My knowledge of Jackson's Json engine is limited, thanks for clearing that up.

Comment (clydet) [reply converter/src/main/java/gov/cms/qpp/conversion/model/validation/MeasureConfig.java]: 👍 @sfradkin

File: converter/src/main/java/gov/cms/qpp/conversion/model/validation/MeasureConfigs.java [hunk 5]
```diff
@@ -8,8 +8,11 @@
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
```
Comment (FlexionSonarQube) [inline converter/src/main/java/gov/cms/qpp/conversion/model/validation/MeasureConfigs.java]: MINOR Remove this unused import 'java.util.concurrent.atomic.AtomicInteger'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: converter/src/main/java/gov/cms/qpp/conversion/validate/CpcQualityMeasureSectionValidator.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	/**
+	 * A holder of CPC+ group specific configuration information.
+	 */
+	enum CpcGroupMinimum {
+		A("outcome", 2),
+		B("complex process", 2);
+
+		private static int NUMBER_OF_MEASURES_REQUIRED = 9;
```
Comment (FlexionSonarQube) [inline converter/src/main/java/gov/cms/qpp/conversion/validate/CpcQualityMeasureSectionValidator.java]: MAJOR Assign this magic number 9 to a well-named constant, and use the constant instead. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS109)
MINOR Rename this field "NUMBER_OF_MEASURES_REQUIRED" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3008)

File: converter/src/main/java/gov/cms/qpp/conversion/validate/Checker.java [hunk 7]
```diff
@@ -332,7 +336,7 @@ public Checker hasMeasures(LocalizedError code, String... measureIds) {
 				return false;
 			}).count();
 
-			if (numberOfMeasuresRequired != numNodesWithWantedMeasureIds) {
+			if (!(numNodesWithWantedMeasureIds >= numberOfMeasuresRequired)) {
```
Comment (FlexionSonarQube) [inline converter/src/main/java/gov/cms/qpp/conversion/validate/Checker.java]: MINOR Use the opposite operator ("<") instead. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1940)

File: converter/src/main/java/gov/cms/qpp/conversion/validate/Checker.java [hunk 8]
```diff
@@ -9,6 +9,7 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Predicate;
```
Comment (FlexionSonarQube) [inline converter/src/main/java/gov/cms/qpp/conversion/validate/Checker.java]: MINOR Remove this unused import 'java.util.function.Predicate'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by FlexionSonarQube** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by clydet** [comment]
@halprin fixed the diff for measures-data.json

**Pull request merged**